library/core/src/option.rs

```diff
@@ -90,6 +90,205 @@
+//! [`filter`]: Option::filter
+//! [`or`]: Option::or
+//! [`or_else`]: Option::or_else
+//! [`xor`]: Option::xor
+//!
+//! ## Iterators
+//!
+//! An [`Option`] can be iterated over. This can be helpful if you need an
+//! iterator that is conditionally empty. The iterator will either produce
+//! a single value (when the [`Option`] is [`Some`]), or produce no values
+//! (when the [`Option`] is [`None`]). For example, [`into_iter`] acts like
+//! [`once(v)`] if the [`Option`] is [`Some(v)`], and like [`empty()`] if
+//! the [`Option`] is [`None`].
+//!
+//! Iterators over [`Option`] come in three types:
+//!
+//! * [`into_iter`] consumes the [`Option`] and produces the contained
+//!   value
+//! * [`iter`] produces an immutable reference of type `&T` to the
+//!   contained value
+//! * [`iter_mut`] produces a mutable reference of type `&mut T` to the
+//!   contained value
+//!
+//! [`Option`] implements the [`FromIterator`] trait, which allows an
+//! iterator over [`Option`] values to be collected into an [`Option`] of a
+//! collection of each contained value of the original [`Option`] values,
+//! or [`None`] if any of the elements was [`None`].
+//!
+//! [`empty()`]: crate::iter::empty
+//! [`FromIterator`]: Option#impl-FromIterator%3COption%3CA%3E%3E
+//! [`into_iter`]: Option::into_iter
+//! [`iter`]: Option::iter
+//! [`iter_mut`]: Option::iter_mut
+//! [`once(v)`]: crate::iter::once
+//! [`Some(v)`]: Some
+//!
+//! An iterator over [`Option`] can be useful when chaining iterators:
+//!
+//! ```
+//! let yep = Some(42);
+//! let nope = None;
+//! let nums: Vec<i32> = (0..4).chain(yep.into_iter()).chain(4..8).collect();
+//! assert_eq!(nums, [0, 1, 2, 3, 42, 4, 5, 6, 7]);
+//! let nums: Vec<i32> = (0..4).chain(nope.into_iter()).chain(4..8).collect();
+//! assert_eq!(nums, [0, 1, 2, 3, 4, 5, 6, 7]);
+//! ```
+//!
+//! One reason to chain iterators in this way is that a function returning
+//! `impl Iterator` must have all possible return values be of the same
+//! concrete type. Chaining an iterated [`Option`] can help with that.
+//!
+//! ```
+//! let yep = Some(42);
+//! let nope = None;
+//!
+//! fn makeiter(opt: Option<i32>) -> impl Iterator<Item = i32> {
+//!     (0..4).chain(opt.into_iter()).chain(4..8)
+//! }
+//! println!("{:?}", makeiter(yep).collect::<Vec<_>>());
+//! println!("{:?}", makeiter(nope).collect::<Vec<_>>());
+//! ```
+//!
+//! If we try to do the same thing, but using pattern matching, we can't
+//! return `impl Iterator` anymore because the concrete types of the return
+//! values differ.
+//!
+//! ```compile_fail,E0308
+//! # use std::iter::{empty, once};
+//! // This won't compile because all possible returns from the function
+//! // must have the same concrete type.
+//! fn makeiter(opt: Option<i32>) -> impl Iterator<Item = i32> {
+//!     match opt {
+//!         Some(x) => return (0..4).chain(once(x)).chain(4..8),
+//!         None => return (0..4).chain(empty()).chain(4..8)
```

Comment:
Yeah, I know it's best practice to omit the `return`. When not using explicit `return`, the type mismatch shows up in the `match` operator instead of with the `impl Iterator` return type, which fails to demonstrate the aspect of the problem that's inherent to `impl Iterator`. Maybe I should add comments to those instances of `return`. I was thinking that in practice, there might be somewhat more involved logic that returns early, so a simple `match` expression wouldn't work.

Reply:
oooooh, that makes perfect sense, thank you